Allow download from lgtm

aeisenberg opened the pull request:
Adds a new command to download databases from lgtm. It's working,
though need to create better error messages on failure.

<!-- Thank you for submitting a pull request. Please read our pull request guidelines before
  submitting your pull request:
  https://github.com/github/vscode-codeql/blob/master/CONTRIBUTING.md#submitting-a-pull-request.
-->

Replace this with a description of the changes your pull request makes.

## Checklist

- [x] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/master/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [x] `@github/product-docs-dsp` has been cc'd in all issues for UI or other user-facing changes made by this pull request.

jcreedcmu commented on extensions/ql-vscode/src/databaseFetcher.ts:
```diff
@@ -1,43 +1,121 @@
+  try {
+    const lgtmUrl = await window.showInputBox({
+      prompt:
+        "Enter the project URL on LGTM (e.g., https://lgtm.com/projects/g/github/codeql)",
```
Any thoughts on why not to require simply the owner/project string, in this example, just `github/codeql`? I feel like that would save a bit of validation complexity.

jcreedcmu replied:
Or do we not have confidence the prefix is always going to be `https://lgtm.com/projects/g/` in the future?

aeisenberg replied:
I was thinking it's just easier for a user to copy the full URL without thinking about which substring to copy and paste it here. And the other thing is that I don't really know what the `g` is.

Maybe I should first get some clarification on what that`g` means.

shati-patel replied:
The `g` means that it's a GitHub project (LGTM also has some [GitLab](https://lgtm.com/projects/gl/sj_vs/lgtm-gitlab-integration-test/) and [BitBucket](https://lgtm.com/projects/b/bitbucket/django_features/) ones.)

jcreedcmu replied:
Hm, that's fair. In that case I might suggest heuristically accepting
`https://lgtm.com/projects/g/github/codeql/?mode=list`
and
`https://lgtm.com/projects/g/github/codeql/alerts`
since the former is what `https://lgtm.com/projects/g/github/codeql/` seems to redirect to that, and `.../alerts` urls are suggested by the lgtm interface in places?

Dunno, not important for this PR necessarily, just a nicety since it's possible to DWIM in these cases.

aeisenberg replied:
I was thinking it's just easier for a user to copy the full URL without thinking about which substring to copy and paste it here. And the other thing is that I don't really know what the `g` is.

Maybe I should first get some clarification on what that`g` means.

jcreedcmu replied:
Ok, so how about just checking that we're at `(www.)lgtm.com`, and there are at least (instead of exactly) 4 components to the path, and the first is `projects`, and we pull out the third and fourth?

aibaars replied:
> Ok, so how about just checking that we're at `(www.)lgtm.com`, and there are at least (instead of exactly) 4 components to the path, and the first is `projects`, and we pull out the third and fourth?

Yes, indeed. Except that you should pull out the second, third and fourth. The URL pattern is `/projects/{provider}/{org}/{name}/sub_pages...`

aibaars commented on extensions/ql-vscode/src/databaseFetcher.ts:
```diff
@@ -218,3 +322,72 @@ async function findDirWithFile(dir: string, ...toFind: string[]): Promise<string
+    }
+
+    const paths = uri.path.split("/").filter((segment) => segment);
+    return paths.length === 4 && paths[0] === "projects" && paths[1] === "g";
```
```suggestion
    return paths.length >= 4 && paths[0] === "projects";
```
The URL pattern is `https://lgtm.com/projects/{provider}/{org}/{name}` . In addition to GitHub.com(`g`), LGTM currently hosts projects from Bitbucket (`b`), GitLab (`gl`) and plain git (`git`). 

After the `{provider}/{org}/{name}` path components, there may be the components related to sub pages. I think it would be best to accept those too.

aeisenberg replied:
Makes sense.

aibaars commented on extensions/ql-vscode/src/databaseFetcher.ts:
```diff
@@ -218,3 +322,72 @@ async function findDirWithFile(dir: string, ...toFind: string[]): Promise<string
+      language,
+    ].join("/")}`;
+  } catch (e) {
+    throw new Error(`Invalid LGTM URL: ${lgtmUrl}`);
```
You should probably include the error message that LGTM returns on error. If a user provides a valid LGTM URL but the wrong language, we would wrongly complain that the LGTM URL is invalid.

aeisenberg replied:
Sure. I can do something, but just note that a user can't request an invalid language since we only allow a user to select from the languages exposed in the project metadata.

aibaars replied:
That's true. Although more precise error message are likely to be helpful in case we ever need to debug a failure.

aeisenberg replied:
Sure. I'll send these messages to the log files.

jcreedcmu commented on extensions/ql-vscode/src/databaseFetcher.ts:
```diff
@@ -339,7 +342,7 @@ function looksLikeLgtmUrl(lgtmUrl: string | undefined): lgtmUrl is string {
 
     const paths = uri.path.split("/").filter((segment) => segment);
-    return paths.length === 4 && paths[0] === "projects" && paths[1] === "g";
+    return paths.length >= 4 && paths[0] === "projects";
```
Maybe just leave a comment in here explaining the URL munging going on --- basically a copy-paste of what @aibaars told us would be good. Otherwise this whole PR lgtm.

aeisenberg replied:
Sure

shati-patel commented:
Minor confusion: The extension lets me select a Python database for https://lgtm.com/projects/g/mrdoob/three.js, but then gives the error `invalid signature: 0x6f63227b`.

Turns out there isn't actually a Python database available: 
![image](https://user-images.githubusercontent.com/42641846/83122207-323ec080-a0cb-11ea-8adc-b067dd7299f1.png)

Is it possible to make this clearer in the logs?

aibaars commented on extensions/ql-vscode/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 - Better formatting and autoindentation when adding QLDoc comments to `.ql` and `.qll` files.
 - Allow for more flexibility when opening a database in the workspace. A user can now choose the actual database folder, or the nested `db-*` folder.
+- Add ability for users to download databases directly from LGTM.
```
Indeed, LGTM.com only.

It would actually work for public projects on LGTM enterprise installations that have anonymous access enabled. This is a rather rare combination, so let's ignore that to avoid confusion ;-)

aibaars commented:
> Minor confusion: The extension lets me select a Python database for https://lgtm.com/projects/g/mrdoob/three.js, but then gives the error `invalid signature: 0x6f63227b`.
> 
> Turns out there isn't actually a Python database available:
> ![image](https://user-images.githubusercontent.com/42641846/83122207-323ec080-a0cb-11ea-8adc-b067dd7299f1.png)
> 
> Is it possible to make this clearer in the logs?

Good point. That project does have python code, but the latest snapshot build failed for python. . As a result the database is not available. Probably best to present the failure returned by the LGTM API. On error there is a JSON response like this
```
{"code":404,"
error":"The specified language could not be found for the project"
}
```

aeisenberg commented:
I think this addresses all comments in the PR so far. Unfortunately, the current download mechanism does not support tar.gz files or other kinds of compression. I'll open a new issue for that.

aeisenberg commented:
See #410.
